gen/cycle.ml

```diff
@@ -927,6 +927,12 @@ let set_same_loc st n0 =
           find_node
             (fun m -> match m.prev.edge.E.edge with
             | E.Fr _|E.Rf _|E.Ws _|E.Leave _|E.Back _
+            | E.Hat|E.Rmw _|E.Irf _|E.Ifr _ -> false
+            | E.Po _|E.Dp _|E.Fenced _|E.Insert _|E.Store|E.Node _ -> true
+            | E.Id -> assert false
+            &&
+            match m.edge.E.edge with
+            | E.Fr _|E.Rf _|E.Ws _|E.Leave _|E.Back _
             | E.Hat|E.Rmw _|E.Irf _|E.Ifr _ -> true
             | E.Po _|E.Dp _|E.Fenced _|E.Insert _|E.Store|E.Node _ -> false
             | E.Id -> assert false) n in
```

Comment:
Can't we do?
```suggestion
            (fun m -> E.is_com m.edge && not (E.is_com m.prev.edge))
```
But as I wrote in the issue you opened, I think the logic should apply if the current edge is com and the next or the previous is not com.

Reply:
Also I just realised that here we consider `Rmw` as a com edge but `is_com` would return `false`. I think `is_com` is correct and here there is a bug, can you confirm this?

Reply:
I think that considering rmw is correct here, as we do not want to allow the case where we start calculating the write values at the end of an rmw node. There is also another issue with the change you suggested, in that it doesn't `assert false` for E.id, which I believe is necessary.

considering these two points, I'm not sure we can make the code much simpler; we can use is_rmw for rmw however we don't have an equivalent for Id, so we might still need to use the pattern matching. However, making such a function is quite trivial, so if you think it makes the code cleaner I can implement this.

In regards to extending the logic to check if the next edge is not com, as mentioned on the github issue, I don't think this would be right as it would allow us to start from the middle of a communication edge chain.